src/wallet/rpcwallet.cpp

```diff
@@ -3149,6 +3149,55 @@ struct StakePeriodRange_T {
 
 typedef std::vector<StakePeriodRange_T> vStakePeriodRange_T;
 
+// Check if we have a Tx that can be counted in staking report
+bool IsTxCountedAsStaked(const CWalletTx* tx)
+{
+    // orphan block or immature
+    if ((!tx->GetDepthInMainChain()) || (tx->GetBlocksToMaturity() > 0))
+        return false;
+
+    // abandoned transactions
+    if (tx->isAbandoned())
+        return false;
+
+    // transaction other than POS block
+    return tx->IsCoinStake();
+}
+
+// Get the amount for a staked tx used in staking report
+CAmount GetTxStakeAmount(const CWalletTx* tx)
+{
+    // use the cached amount if available
+    if ((tx->fCreditCached || tx->fColdStakingCreditCached) && (tx->fDebitCached || tx->fColdStakingDebitCached))
+        return tx->nCreditCached + tx->nColdStakingCreditCached - tx->nDebitCached - tx->nColdStakingDebitCached;
+    // Check for cold staking 
+    else if (tx->vout[1].scriptPubKey.IsColdStaking())
+        return tx->GetCredit(pwalletMain->IsMine(tx->vout[1])) - tx->GetDebit(pwalletMain->IsMine(tx->vout[1]));
+
+    return tx->GetCredit(ISMINE_SPENDABLE) + tx->GetCredit(ISMINE_STAKABLE) - tx->GetDebit(ISMINE_SPENDABLE) - tx->GetDebit(ISMINE_STAKABLE);
```

Comment:
ignoring ISMINE_SPENDABLE_STAKABLE would make this to fail for wallets holding both the spending and staking key (not an ideal scenario from a security perspective but its technically possible)

Reply:
I'll see if I can add more scenarios to the test case

Reply:
I think this scenario was already covered by the test cases

Please check this code in the test and let me know what you think:
```python
        # Import the 2 keys into a third wallet
        self.nodes[0].importprivkey(spending_address_private_key)
        self.nodes[0].importprivkey(staking_address_private_key)
```

Basically in the test cases, there are 3 modes, node 1 has spending key, node 2 has staking key and node 0 has both keys imported

Reply:
yes you are right, my bad as `ISMINE_SPENDABLE_STAKABLE = ISMINE_SPENDABLE | ISMINE_STAKABLE`

Reply:
I forget why, but in the 4.6.0 fix that I created for this

This would not get the correct amounts:
```cpp
tx->GetCredit(ISMINE_SPENDABLE_STAKABLE) - tx->GetDebit(ISMINE_SPENDABLE_STAKABLE);
```

Bu this would:
```cpp
tx->GetCredit(ISMINE_SPENDABLE) + tx->GetCredit(ISMINE_STAKABLE) - tx->GetDebit(ISMINE_SPENDABLE) - tx->GetDebit(ISMINE_STAKABLE);
```

While in theory they should be the same, but I don't remember exactly why.

I think there was code in `GetCredit` and `GetDebit` that was causing an issue when passing `ISMINE_SPENDABLE_STAKABLE`

Reply:
`ISMINE_SPENDABLE_STAKABLE` is Spendable+Stakable

Reply:
Anyways, I think the scenarios are covered in the test case

And I also added the time travel stuff that we discussed with @craigmacgregor during the 4.6.0 release which we skipped due to time constraints.